### PR TITLE
Minor Development Changes

### PR DESCRIPTION
**Pull request by bennyman123abc**
This pull request changes the following things in both the frontend and the backend:

### Frontend
- Finished the dark mode to match toolbar and footer
- Added a button to refresh your tag without editing in the Account page
- Add @HEYimHeroic to the credits of the project for their work on Miis
- Made the credits page component based
- Added a button to play Mario credits music (From YouTube) to the credits page

### Backend
- Modified the tag renderer to make use of the background fields in avatar and mii in overlay files, if present
- Allowed the alignment of friend codes to fix overlay7

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -45,23 +45,24 @@
...
     "pino": "7.6.4",
     "pino-pretty": "7.5.0",
     "prop-types": "15.8.1",
     "react": "17.0.2",
     "react-bootstrap": "2.1.1",
     "react-dom": "17.0.2",
     "react-paginate": "8.1.0",
-    "safe-json-stringify": "1.2.0",
     "react-toastify": "8.1.0",
+    "react-youtube": "^7.14.0",
```
Comment (Brawl345) [inline package.json:64]: Please use just the version number instead of `^` to avoid breaking updates.

File: src/components/account/PrivateKey.jsx [hunk 2]
```diff
@@ -33,21 +34,24 @@ function PrivateKey({ randkey, toggleModal }) {
...
+          {/* Copying to clipboard does NOT work on HTTP, only HTTPS */}
+          { ENV.BASE_URL.startsWith("https://") ? 
           <Button variant={copySuccess ? 'success' : 'light'} onClick={copy}>
             <FontAwesomeIcon
               className="me-1"
               icon={copySuccess ? faCheck : faCopy}
             />
             {copySuccess ? 'Copied!' : 'Copy'}
           </Button>
+          : <></> }
```
Comment (Brawl345) [inline src/components/account/PrivateKey.jsx:54]: You can return `null` instead of an empty fragment. You can shorten this to 

```jsx
{ENV.BASE_URL.startsWith("https://") &&
(<Button variant={copySuccess ? 'success' : 'light'} onClick={copy}>
            <FontAwesomeIcon
              className="me-1"
              icon={copySuccess ? faCheck : faCopy}
            />
            {copySuccess ? 'Copied!' : 'Copy'}
          </Button>)
}
```

File: src/components/account/RefreshTagButton.jsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { toast } from 'react-toastify';
+import { Button } from 'react-bootstrap';
+
+export default function RefreshTagButton() {
+    const refresh = async () => {
+        const response = await fetch('/api/account/refresh-tag', {
+            method: 'GET',
```
Comment (Brawl345) [inline src/components/account/RefreshTagButton.jsx:7]: Use `POST` to avoid CSRF

File: src/components/credits/Contributor.jsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import ExternalLink from "@/components/shared/ExternalLink";
+
+export default function Contributor(props) {
```
Comment (Brawl345) [inline src/components/credits/Contributor.jsx:3]: Please use props destructuring instead: https://medium.com/@lcriswell/destructuring-props-in-react-b1c295005ce0

It helps with typing and you can (should) add prop-types: https://reactjs.org/docs/typechecking-with-proptypes.html#function-components

File: src/components/credits/Contributor.jsx [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    return (
+        <li>
+            <strong>
+                {props.link ? <NameWithLink name={props.name} link={props.link} /> : props.name}
+            </strong>: {props.children}
+        </li>
+    );
+}
+
+function NameWithLink(props) {
```
Comment (Brawl345) [inline src/components/credits/Contributor.jsx:13]: Same as above and move it above Contributor or better, move it to a seperate file (so one JSX = one component)

File: src/components/credits/CreditsMusic.jsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
+import YouTube from 'react-youtube';
+import { Button } from 'react-bootstrap';
+import React from 'react';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faPlay, faStop } from '@fortawesome/free-solid-svg-icons';
+import { toast } from 'react-toastify';
+
+export default class CreditsMusic extends React.Component {
```
Comment (Brawl345) [inline src/components/credits/CreditsMusic.jsx:8]: Please use function components instead of class components: https://reactjs.org/docs/components-and-props.html#function-and-class-components

This helps avoid errors with `this`, makes the codebase more consistent (everything is a function component) and makes it possible to use hooks.

Comment (Brawl345) [reply src/components/credits/CreditsMusic.jsx:8]: Still thinking if we should really use YouTube, because it leads to privacy issues and I don't know if react-youtube uses the `youtube-nocookie` domain (which is required by GDPR). Maybe we should just upload the music ourselves...?

File: src/components/shared/ExternalLink.jsx [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export default function ExternalLink(props) {
```
Comment (Brawl345) [inline src/components/shared/ExternalLink.jsx:1]: Use props destructuring (see other comment on Contributor.jsx)

File: src/pages/api/account/refresh-tag.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        return response
+            .status(HTTP_CODE.INTERNAL_SERVER_ERROR)
+            .json({ error: "Error while creating tag." 
+        });
+    }
+
+    return response.status(HTTP_CODE.OK).json({ success: true });
+}
+
+const handler = ncWithSession().get(refreshTag);
```
Comment (Brawl345) [inline src/pages/api/account/refresh-tag.js:34]: Use `post` to avoid CSRF

File: src/pages/api/account/refresh-tag.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
+import HTTP_CODE from "@/lib/constants/httpStatusCodes";
+import { makeBanner } from "@/lib/riitag/banner";
+import logger from "@/lib/logger";
+import { ncWithSession } from "@/lib/routing";
+
+async function refreshTag(request, response) {
```
Comment (Brawl345) [inline src/pages/api/account/refresh-tag.js:6]: Maybe we should rate-limit this? You can see how it's done in the update-tag stuff. You can use the username as a key before even hitting the database.

File: src/pages/credits.jsx [hunk 10]
```diff
@@ -3,8 +3,56 @@ import Image from 'next/image';
...
+  const accountInfo = await prisma.user.findFirst({
+    where: {
+      username,
+    },
+    select: {
+      name_on_riitag: true,
+    },
+  });
+
+  return { props: { accountInfo: JSON.parse(safeJsonStringify(accountInfo)) } };
```
Comment (Brawl345) [inline src/pages/credits.jsx:52]: Dou you need `safeJsonStringify` here? You could just use `accountinfo.name_on_riitag` directly. You can then use a fallback if no name was set.

File: src/pages/credits.jsx [hunk 11]
```diff
@@ -3,8 +3,56 @@ import Image from 'next/image';
...
+});
+
+function CreditsPage({ accountInfo }) {
```
Comment (Brawl345) [inline src/pages/credits.jsx:55]: Same as above - just passing `name_on_riitag` is enough.

File: package.json (hunk 1 above)
Comment (bennyman123abc) [reply package.json:65]: Ah yes, I didn't add it manually; just with npm i

File: src/components/credits/Contributor.jsx (hunk 5 above)
Comment (bennyman123abc) [reply src/components/credits/Contributor.jsx:13]: I put it in that file because there's only one insurance of the component being used

File: src/pages/credits.jsx (hunk 10 above)
Comment (bennyman123abc) [reply src/pages/credits.jsx:52]: I just copied the code from another file here and forgot to clean it up

File: src/components/credits/Contributor.jsx (hunk 5 above)
Comment (Brawl345) [reply src/components/credits/Contributor.jsx:13]: That's totally fine but I would still split them.

File: src/pages/api/account/refresh-tag.js (hunk 8 above)
Comment (bennyman123abc) [reply src/pages/api/account/refresh-tag.js:34]: I'm unsure what you mean here?

Comment (Brawl345) [reply src/pages/api/account/refresh-tag.js:34]: You should change the `get` to `post` together with "`method:`" where this URL is called (I commented on it in the other file). This is so that an attacker can't just do actions on behalf of the user.

File: src/components/credits/CreditsMusic.jsx (hunk 6 above)
Comment (bennyman123abc) [reply src/components/credits/CreditsMusic.jsx:8]: Uploading the music ourselves would potentially cause copyright issues, which is why I decided to use a hidden YouTube player.

Comment (bennyman123abc) [reply src/components/credits/CreditsMusic.jsx:8]: Also, it uses the proper YouTube player for your region. If nothing else, we can include a cookie notice on the front page as we use cookies ourselves for authentication.

Comment (Brawl345) [reply src/components/credits/CreditsMusic.jsx:8]: > Uploading the music ourselves would potentially cause copyright issues, which is why I decided to use a hidden YouTube player.

Hm that's true. I can also take a look at it, but not before the weekend though.

> If nothing else, we can include a cookie notice on the front page

I would rather avoid that since it needs opt-in. Cookies for authentication is fine since they are functional cookies which need no opt-in.

**Comment by Brawl345** [comment]
(You can ignore the Code Style stuff from Prettier, I can also clean it up)

**Comment by Brawl345** [comment]
I cleaned it up a bit and removed the credits music for now. Sorry for "ruining the fun", but due to privacy concerns, this is not doable in this way. We can revisit this in the future (maybe with a "when clicked this will contact youtube" or something like that before the button becomes available on the credits page).
